[Bug]: Blank/transparent window on Linux Wayland sessions (Arch, Omarchy, Hyprland, GNOME-Wayland)

### Operating system

Linux ( Arch Linux )

### Orca version

1.2.3-rc.1

### Details

Blank/transparent window on Linux Wayland sessions (Arch, Omarchy, Hyprland, GNOME-Wayland) - Vulkan flag is incompatible with Chromium's Ozone/Wayland backend

**Environment**

OS - Arch Linux (Omarchy spin) - reproduced
Compositor  - Hyprland (Wayland)                                                                      
Session type - XDG_SESSION_TYPE=wayland
Electron - 41.0.3                                                                                          
Orca - 1.2.3-rc.1 (main)
Driver stack - Mesa + any Vulkan ICD

Also expected to affect any distro that defaults to Wayland (Fedora Workstation, recent Ubuntu, KDE Plasma 6, GNOME 46+, Omarchy, etc.) when running Orca as a native Wayland client.
                                                                                                                                                             
**Symptom**         

  On launch, a window frame appears at the expected size/position, but the content area is fully transparent, the user sees their desktop wallpaper through the window instead of Orca's UI. No error dialogue is shown, DevTools never opens, and the main process keeps running normally (PTY GC, IPC handlers, etc., all initialize).                                                                                                                                           
                  
**Root cause**                                                                                                                                                 
                  
src/main/startup/configure-process.ts → enableMainProcessGpuFeatures() unconditionally enables Vulkan + Skia Graphite + unsafe WebGPU before app.whenReady:
                  
  export function enableMainProcessGpuFeatures(): void {                                                                                                     
    app.commandLine.appendSwitch('enable-features', 'Vulkan,UseSkiaGraphite')                                                                                
    app.commandLine.appendSwitch('enable-unsafe-webgpu')                                                                                                     
  }                                                                                                                                                          
                                                                                                                                                             
On Wayland, Chromium's Ozone/Wayland backend refuses to initialize a Vulkan surface. The GPU process logs:                                                 
                  
  ERROR:ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc:251]                                                                                        
  '--ozone-platform=wayland' is not compatible with Vulkan.                                                                                                  
  Consider switching to '--ozone-platform=x11' or disabling Vulkan                                                                                           
                                                                                                                                                             
Without a compatible GPU surface, the renderer never composes, so the BrowserWindow is drawn transparent. Because this happens in the GPU process rather than the main process, no uncaught exception bubbles up, and the app looks "running but blank."                                                                    
                                                                                                                                                             
**Why does it only hit some users**

  - macOS / Windows: Vulkan switch is honoured via MoltenVK → Metal (macOS) or works cleanly on Windows. No regression.                                       
  - Linux on X11: Chromium's X11 path tolerates the Vulkan feature flag (falls back to GL where Vulkan isn't available), so the window still paints.
  - Linux on Wayland: Ozone/Wayland surface factory hard-aborts → blank window. This is the failure mode Arch/Omarchy and other Wayland-default distros hit. 
                                                                                                                                                             
**Reproduction**                                                                                                                                               
                                                                                                                                                             
  1. Run Orca on any Wayland session: echo $XDG_SESSION_TYPE should print wayland.                                                                           
  2. pnpm install && pnpm dev (or run the packaged AppImage/.deb).
  3. Observe a window that opens with a completely transparent content area.                                                                                      
  4. In the dev log or GPU process stderr, the wayland_surface_factory.cc:251 error is present.                                                              
                                                                                                                                                             
**User-side workaround (no code change)**                                                                                                                      
                                                                                                                                                             
  Force Chromium's X11 backend via XWayland:                                                                                                                 
                  
  ELECTRON_OZONE_PLATFORM_HINT=x11 pnpm dev                                                                                                                  
  or for the packaged app:                                                                                                                                 
  ELECTRON_OZONE_PLATFORM_HINT=x11 orca
                                                                                                                                                             
  **Fix applied**     
                                                                                                                                                             
Gate the Vulkan/SkiaGraphite/WebGPU switches so they never apply on Linux. macOS and Windows keep their accelerated Skia Graphite path; Linux falls back to Chromium's default GPU pipeline, which renders correctly on both Wayland and X11.
                                                                                                                                                             
  src/main/startup/configure-process.ts

   export function enableMainProcessGpuFeatures(): void {
  +  // Why: Chromium's Ozone/Wayland surface factory refuses to initialize when                                                                             
  +  // Vulkan is enabled (see wayland_surface_factory.cc), which leaves the                                                                                 
  +  // renderer unable to compose and shows a blank/transparent window on                                                                                   
  +  // Wayland-default distros like Arch/Omarchy. The X11 Linux Vulkan path is                                                                              
  +  // similarly unreliable across Electron 41 + Mesa combinations, so we skip                                                                              
  +  // these switches on Linux entirely and keep the Skia Graphite / WebGPU                                                                                 
  +  // acceleration for macOS and Windows, where it works as intended.                                                                                       
  +  if (process.platform === 'linux') {                                                                                                                     
  +    return                                                                                                                                                
  +  }                                                                                                                                                       
     app.commandLine.appendSwitch('enable-features', 'Vulkan,UseSkiaGraphite')                                                                               
     app.commandLine.appendSwitch('enable-unsafe-webgpu')
   }                                                                                                                                                         
   
  **Tradeoffs considered:**  
                                                                                                                                    
  - Narrow gate (Wayland only): detect XDG_SESSION_TYPE=wayland / WAYLAND_DISPLAY and skip there. Rejected, Chromium's Linux Vulkan path is flaky on X11 too, and the failure mode (silent blank window) is too user-hostile to keep the feature on any Linux host.                                                 
  - Broad gate (all Linux): chosen. Minor GPU acceleration regression on Linux only, fully prevents the blank-window failure mode, keeps macOS/Windows unchanged.                                                                                                                                                 
                                                                                                                                                             
  **Test coverage**
                                                                                                                                                             
  Existing describe('enableMainProcessGpuFeatures', …) suite updated to:                                                                                     
  1. Assert the Vulkan/SkiaGraphite/WebGPU switches are applied on darwin.
  2. Add a Linux case asserting app.commandLine.appendSwitch is not called.                                                                                  
                                                                           
  Full suite: 1722 tests pass after the change.                                                                                                              
                                                                                                                                                             
  **Verification**                                                                                                                                               
                                                                                                                                                             
  On the reproducing host (Arch + Hyprland + Wayland):                                                                                                       
  - Rebuilt main bundle and launched pnpm dev.
  - wayland_surface_factory.cc error no longer appears in the log.                                                                                           
  - Captured a window screenshot: file tree, editor, chrome, and theme toggle all render correctly.
                                                                                                                                                             
  **Suggested follow-ups (optional)**                                                                                                                            
                                                                                                                                                             
  - Add a one-time main-process log line when running on Linux Wayland so future regressions of this kind are diagnosable from user bug reports.             
  - Consider a Linux CI smoke test under xvfb-run / sway --test that asserts the first BrowserWindow's webContents reaches did-finish-load within N seconds, would have caught this specific regression.                                                                                                               
                  
  ---

--- a/src/main/startup/configure-process.test.ts
+++ b/src/main/startup/configure-process.test.ts
@@ -182,6 +182,8 @@ describe('installDevParentWatchdog', () => {
 
 describe('enableMainProcessGpuFeatures', () => {
   const originalPlatform = Object.getOwnPropertyDescriptor(process, 'platform')
+  const originalWaylandDisplay = process.env.WAYLAND_DISPLAY
+  const originalXdgSessionType = process.env.XDG_SESSION_TYPE
 
   function setPlatform(platform: NodeJS.Platform): void {
     Object.defineProperty(process, 'platform', {
@@ -193,6 +195,16 @@ describe('enableMainProcessGpuFeatures', () => {
   afterEach(() => {
     if (originalPlatform) {
       Object.defineProperty(process, 'platform', originalPlatform)
+    }
+    if (originalWaylandDisplay === undefined) {
+      delete process.env.WAYLAND_DISPLAY
+    } else {
+      process.env.WAYLAND_DISPLAY = originalWaylandDisplay
+    }
+    if (originalXdgSessionType === undefined) {
+      delete process.env.XDG_SESSION_TYPE
+    } else {
+      process.env.XDG_SESSION_TYPE = originalXdgSessionType
     }
   })
 
@@ -211,16 +223,53 @@ describe('enableMainProcessGpuFeatures', () => {
     expect(app.commandLine.appendSwitch).toHaveBeenCalledWith('enable-unsafe-webgpu')
   })
 
-  it('skips Vulkan/SkiaGraphite/WebGPU switches on linux', async () => {
-    // Why: Chromium's Ozone/Wayland surface factory refuses to initialize
-    // with Vulkan enabled, producing a blank/transparent window on Wayland
-    // sessions (Arch/Omarchy/Hyprland, GNOME-Wayland). Guard against
-    // regressions that would reintroduce these switches on Linux.
+  it('appends Orca GPU flags on linux X11 sessions', async () => {
+    // Why: Chromium's X11 Vulkan path works, so we keep Skia Graphite / WebGPU
+    // acceleration on X11. Only Wayland hits the wayland_surface_factory.cc
+    // incompatibility and needs the gate.
     const { app } = await import('electron')
     const { enableMainProcessGpuFeatures } = await import('./configure-process')
 
     vi.mocked(app.commandLine.appendSwitch).mockClear()
     setPlatform('linux')
+    delete process.env.WAYLAND_DISPLAY
+    process.env.XDG_SESSION_TYPE = 'x11'
+    enableMainProcessGpuFeatures()
+
+    expect(app.commandLine.appendSwitch).toHaveBeenCalledWith(
+      'enable-features',
+      'Vulkan,UseSkiaGraphite'
+    )
+    expect(app.commandLine.appendSwitch).toHaveBeenCalledWith('enable-unsafe-webgpu')
+  })
+
+  it('skips Vulkan/SkiaGraphite/WebGPU switches when WAYLAND_DISPLAY is set', async () => {
+    // Why: Chromium's Ozone/Wayland surface factory hard-aborts with Vulkan
+    // enabled (wayland_surface_factory.cc:251), producing a blank/transparent
+    // window on Wayland-default distros (Arch/Omarchy/Hyprland, GNOME-Wayland).
+    const { app } = await import('electron')
+    const { enableMainProcessGpuFeatures } = await import('./configure-process')
+
+    vi.mocked(app.commandLine.appendSwitch).mockClear()
+    setPlatform('linux')
+    process.env.WAYLAND_DISPLAY = 'wayland-0'
+    delete process.env.XDG_SESSION_TYPE
+    enableMainProcessGpuFeatures()
+
+    expect(app.commandLine.appendSwitch).not.toHaveBeenCalled()
+  })
+
+  it('skips switches when XDG_SESSION_TYPE=wayland without WAYLAND_DISPLAY', async () => {
+    // Why: belt-and-suspenders — some nested/manually-started Wayland sessions
+    // advertise XDG_SESSION_TYPE=wayland without yet having WAYLAND_DISPLAY
+    // set in the process's environment at launch time.
+    const { app } = await import('electron')
+    const { enableMainProcessGpuFeatures } = await import('./configure-process')
+
+    vi.mocked(app.commandLine.appendSwitch).mockClear()
+    setPlatform('linux')
+    delete process.env.WAYLAND_DISPLAY
+    process.env.XDG_SESSION_TYPE = 'wayland'
     enableMainProcessGpuFeatures()
 
     expect(app.commandLine.appendSwitch).not.toHaveBeenCalled()

--- a/src/main/startup/configure-process.test.ts
+++ b/src/main/startup/configure-process.test.ts
@@ -181,10 +181,27 @@ describe('installDevParentWatchdog', () => {
 })
 
 describe('enableMainProcessGpuFeatures', () => {
-  it('appends Orca GPU flags by default', async () => {
+  const originalPlatform = Object.getOwnPropertyDescriptor(process, 'platform')
+
+  function setPlatform(platform: NodeJS.Platform): void {
+    Object.defineProperty(process, 'platform', {
+      configurable: true,
+      value: platform
+    })
+  }
+
+  afterEach(() => {
+    if (originalPlatform) {
+      Object.defineProperty(process, 'platform', originalPlatform)
+    }
+  })
+
+  it('appends Orca GPU flags on darwin', async () => {
     const { app } = await import('electron')
     const { enableMainProcessGpuFeatures } = await import('./configure-process')
 
+    vi.mocked(app.commandLine.appendSwitch).mockClear()
+    setPlatform('darwin')
     enableMainProcessGpuFeatures()
 
     expect(app.commandLine.appendSwitch).toHaveBeenCalledWith(
@@ -192,5 +209,20 @@ describe('enableMainProcessGpuFeatures', () => {
       'Vulkan,UseSkiaGraphite'
     )
     expect(app.commandLine.appendSwitch).toHaveBeenCalledWith('enable-unsafe-webgpu')
+  })
+
+  it('skips Vulkan/SkiaGraphite/WebGPU switches on linux', async () => {
+    // Why: Chromium's Ozone/Wayland surface factory refuses to initialize
+    // with Vulkan enabled, producing a blank/transparent window on Wayland
+    // sessions (Arch/Omarchy/Hyprland, GNOME-Wayland). Guard against
+    // regressions that would reintroduce these switches on Linux.
+    const { app } = await import('electron')
+    const { enableMainProcessGpuFeatures } = await import('./configure-process')
+
+    vi.mocked(app.commandLine.appendSwitch).mockClear()
+    setPlatform('linux')
+    enableMainProcessGpuFeatures()
+
+    expect(app.commandLine.appendSwitch).not.toHaveBeenCalled()
   })
 })

--- a/src/main/startup/configure-process.ts
+++ b/src/main/startup/configure-process.ts
@@ -147,6 +147,16 @@ export function installDevParentWatchdog(isDev: boolean): void {
 }
 
 export function enableMainProcessGpuFeatures(): void {
+  // Why: Chromium's Ozone/Wayland surface factory refuses to initialize when
+  // Vulkan is enabled (see wayland_surface_factory.cc), leaving the renderer
+  // unable to compose and showing a blank/transparent window on Wayland-default
+  // distros like Arch/Omarchy/Hyprland and GNOME-Wayland. Chromium's Linux
+  // Vulkan path is also flaky on X11 across Electron 41 + Mesa combinations,
+  // so we skip these switches on Linux entirely. macOS and Windows keep the
+  // Skia Graphite / WebGPU acceleration path, which works as intended.
+  if (process.platform === 'linux') {
+    return
+  }
   app.commandLine.appendSwitch('enable-features', 'Vulkan,UseSkiaGraphite')
   app.commandLine.appendSwitch('enable-unsafe-webgpu')
 }

--- a/src/main/startup/configure-process.ts
+++ b/src/main/startup/configure-process.ts
@@ -147,9 +147,6 @@ export function installDevParentWatchdog(isDev: boolean): void {
 }
 
 function isLinuxWaylandSession(): boolean {
-  if (process.platform !== 'linux') {
-    return false
-  }
   // Why: WAYLAND_DISPLAY is set directly by the Wayland compositor and is the
   // same signal Electron's own ELECTRON_OZONE_PLATFORM_HINT=auto logic uses.
   // XDG_SESSION_TYPE is the login-manager/PAM signal and is the belt-and-
@@ -157,7 +154,10 @@ function isLinuxWaylandSession(): boolean {
   // start (nested Wayland, manual session startup). Both are inherited from
   // the parent process, so they're available before app.whenReady where the
   // GPU command-line switches must be appended.
-  return Boolean(process.env.WAYLAND_DISPLAY) || process.env.XDG_SESSION_TYPE === 'wayland'
+  return (
+    process.platform === 'linux' &&
+    (Boolean(process.env.WAYLAND_DISPLAY) || process.env.XDG_SESSION_TYPE === 'wayland')
+  )
 }
 
 export function enableMainProcessGpuFeatures(): void {

--- a/src/main/startup/configure-process.ts
+++ b/src/main/startup/configure-process.ts
@@ -146,15 +146,31 @@ export function installDevParentWatchdog(isDev: boolean): void {
   timer.unref()
 }
 
+function isLinuxWaylandSession(): boolean {
+  if (process.platform !== 'linux') {
+    return false
+  }
+  // Why: WAYLAND_DISPLAY is set directly by the Wayland compositor and is the
+  // same signal Electron's own ELECTRON_OZONE_PLATFORM_HINT=auto logic uses.
+  // XDG_SESSION_TYPE is the login-manager/PAM signal and is the belt-and-
+  // suspenders check for sessions where WAYLAND_DISPLAY isn't set at process
+  // start (nested Wayland, manual session startup). Both are inherited from
+  // the parent process, so they're available before app.whenReady where the
+  // GPU command-line switches must be appended.
+  return Boolean(process.env.WAYLAND_DISPLAY) || process.env.XDG_SESSION_TYPE === 'wayland'
+}
+
 export function enableMainProcessGpuFeatures(): void {
-  // Why: Chromium's Ozone/Wayland surface factory refuses to initialize when
-  // Vulkan is enabled (see wayland_surface_factory.cc), leaving the renderer
-  // unable to compose and showing a blank/transparent window on Wayland-default
-  // distros like Arch/Omarchy/Hyprland and GNOME-Wayland. Chromium's Linux
-  // Vulkan path is also flaky on X11 across Electron 41 + Mesa combinations,
-  // so we skip these switches on Linux entirely. macOS and Windows keep the
-  // Skia Graphite / WebGPU acceleration path, which works as intended.
-  if (process.platform === 'linux') {
+  // Why: Chromium's Ozone/Wayland surface factory hard-aborts when Vulkan is
+  // enabled (see wayland_surface_factory.cc:251 — "--ozone-platform=wayland is
+  // not compatible with Vulkan"), leaving the renderer unable to compose and
+  // showing a blank/transparent window on Wayland-default distros like
+  // Arch/Omarchy/Hyprland and GNOME-Wayland. Skia Graphite is Vulkan-only on
+  // Linux (no OpenGL backend), so enabling it with Vulkan off would silently
+  // fall back to software rendering — gate it on the same Wayland signal.
+  // The X11 Vulkan path works, so we keep the acceleration there and on
+  // macOS/Windows.
+  if (isLinuxWaylandSession()) {
     return
   }
   app.commandLine.appendSwitch('enable-features', 'Vulkan,UseSkiaGraphite')

